app/ore/project/factory/ProjectFactory.scala

```diff
@@ -103,32 +99,45 @@ trait ProjectFactory {
 
     // create and load a new PluginFile instance for further processing
     val plugin = new PluginFile(pluginPath, sigPath, owner)
-    plugin.loadMeta()
-    plugin
+    val result = plugin.loadMeta()
+    result match {
+      case Left(_) => Left(plugin)
+      case Right(errorMessage) => Right(errorMessage)
+    }
   }
 
   def processSubsequentPluginUpload(uploadData: PluginUpload,
                                     owner: User,
                                     project: Project)(implicit ec: ExecutionContext, messages: Messages): EitherT[Future, String, PendingVersion] = {
-    val plugin = this.processPluginUpload(uploadData, owner)
-    if (!plugin.meta.get.getId.equals(project.pluginId))
-      EitherT.leftT("error.version.invalidPluginId")
-    else {
-      EitherT(
-        for {
-          (channels, settings) <- (project.channels.all, project.settings).parTupled
-          version = this.startVersion(plugin, project, settings, channels.head.name)
-          modelExists <- version.underlying.exists
-        } yield {
-          if (modelExists && this.config.projects.get[Boolean]("file-validate"))
-            Left("error.version.duplicate")
-          else {
-            version.cache()
-            Right(version)
+    this.processPluginUpload(uploadData, owner) match {
+      case Left(plugin) => if (!plugin.data.flatMap(_.getId).contains(project.pluginId))
+        EitherT.leftT("error.version.invalidPluginId")
+      else {
+        EitherT(
+          for {
+            (channels, settings) <- (project.channels.all, project.settings).parTupled
+            version = this.startVersion(plugin, project, settings, channels.head.name)
+            modelExists <- version match {
+              case Left(v) => v.underlying.exists
+              case Right(_) => Future.successful(false)
+            }
+          } yield {
+            version match {
+              case Left(v) => if (modelExists && this.config.projects.get[Boolean]("file-validate"))
```

Comment:
Place this if in the match clause, and make a new clause if the test doesn't hold